### PR TITLE
Separate stream extensions for filling buffers and getting data

### PR DESCRIPTION
**Pull request by JamesNK**
To prevent misusing extension methods I have separate methods for filling the passed in buffer and getting data.

ReadAtLeastLengthAsync and ReadUntilEndAsync could be renamed to be prefixed with Get, Fetch or Load if people don't like "Read" extension methods on Stream not returning a length.

### CODE REVIEW

File: src/Servers/Kestrel/shared/test/TestApp.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public static async Task EchoApp(HttpContext httpContext)
 
             if (buffer.Length > 0)
             {
-                await request.Body.ReadUntilEndAsync(buffer).DefaultTimeout();
+                await request.Body.FillBufferUntilEndAsync(buffer).DefaultTimeout();
```
Comment (halter73) [inline src/Servers/Kestrel/shared/test/TestApp.cs:22]: We should keep verifying we've reached the end-of-stream in these cases.

Comment (halter73) [reply src/Servers/Kestrel/shared/test/TestApp.cs:22]: I need to read more closely. It's FillBuffer*UntilEnd*Async

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs [hunk 2]
```diff
@@ -1725,7 +1725,7 @@ public async Task SynchronousReadsCanBeDisallowedGlobally()
                 Assert.Equal(CoreStrings.SynchronousReadsDisallowed, ioEx2.Message);
 
                 var buffer = new byte[5];
-                var length = await context.Request.Body.ReadUntilEndAsync(buffer).DefaultTimeout();
+                var length = await context.Request.Body.FillEntireBufferAsync(buffer).DefaultTimeout();
```
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs:1728]:
```suggestion
                var length = await context.Request.Body.FillEntireBufferUntilEndAsync(buffer).DefaultTimeout();
```

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs [hunk 3]
```diff
@@ -1725,7 +1725,7 @@ public async Task SynchronousReadsCanBeDisallowedGlobally()
                 Assert.Equal(CoreStrings.SynchronousReadsDisallowed, ioEx2.Message);
 
                 var buffer = new byte[5];
-                var length = await context.Request.Body.ReadUntilEndAsync(buffer).DefaultTimeout();
+                var length = await context.Request.Body.FillEntireBufferUntilEndAsync(buffer).DefaultTimeout();
```
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs:1728]: I think I messed up my last suggestion. I'll just apply it myself this time.

```suggestion
                var length = await context.Request.Body.FillBufferUntilEndAsync(buffer).DefaultTimeout();
```

File: src/Servers/Kestrel/shared/test/StreamExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (read != 0)
+            {
+                Assert.Equal(0, await stream.ReadAsync(new byte[1], 0, 1, cancellationToken));
+            }
+
+            return offset;
+        }
+
+        public static async Task<int> FillEntireBufferAsync(this Stream stream, byte[] buffer, CancellationToken cancellationToken = default)
```
Comment (halter73) [inline src/Servers/Kestrel/shared/test/StreamExtensions.cs:34]: Does anything call this now?

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/shared/test/StreamExtensions.cs (hunk 4 above)
Comment (JamesNK) [reply src/Servers/Kestrel/shared/test/StreamExtensions.cs:34]: image

Comment (halter73) [reply src/Servers/Kestrel/shared/test/StreamExtensions.cs:34]: [Yep](https://github.com/dotnet/aspnetcore/blob/jamesnk/streams-readmethods/src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs#L1728).

**Pull request merged**